stdpopsim/models.py

```diff
@@ -137,6 +137,176 @@ def asdict(self):
+                P[index_array[x, y], index_array[x, y]] = K_x * (K_y - self._delta(x, y))
+        P = P / np.sum(P)
+        r = np.zeros(num_steps)
+        steps = np.linspace(0, end, num_steps)
+        dt = steps[1] - steps[0]
+        for j in range(num_steps):
+            N_M = self._get_N_M_at_t(steps[j])
+            pop_sizes = N_M["population_sizes"]
+            mig_matrix = N_M["migration_matrix"]
+            C = self._get_C_from_N(pop_sizes)
+            Mp = self._get_Mp_from_M(mig_matrix)
+            G = self._get_G_from_Mp_C(Mp, C)
+            P = P * np.exp(dt * G)
+            r[j] = np.sum(np.matmul(P, C)) / np.sum(P)
+        return steps, r
+
+    def num_pops(self):
+        """
+        This function returns the number populations
+        defined by the demographic model
+        """
+        ddb = msprime.DemographyDebugger(**self.asdict())
+        return len(ddb.epochs[0].populations)
+
+    # TODO Rename
+    def _get_epoch_at_t(self, t):
+        """
+        Given a time, t (in generations), find and return
+        the Epoch for which this belongs.
+        If there is no epoch that exists for the given t
+        then it returns None.
+
+        <class 'int'> -> <class 'msprime.simulations.Epoch'>
+        """
+
+        ddb = msprime.DemographyDebugger(**self.asdict())
+        epochs = ddb.epochs
+        j = 0
+        while epochs[j].end_time <= t:
+            j += 1
+        return j, epochs[j]
+
+    # TODO Rename
+    def _get_N_M_at_t(self, t):
+        """
+        Given a time, t (in generations), find and return
+        a dictionary containing:
+        1: "population_sizes" the vector N which should represent effective
+        population size for each populations at time t and,
+        2: "migration_matrix" The migration matrix for for the populations at time t
+
+        <class 'int'> -> <class 'dict'>
+        """
+
+        epochIndex, epoch = self._get_epoch_at_t(t)
+        ddb = msprime.DemographyDebugger(**self.asdict())
+        N = ddb.population_size_history[:, epochIndex]
+        for i, pop in enumerate(epoch.populations):
+            s = t - epoch.start_time
+            g = pop.growth_rate
+            N[i] *= np.exp(-1 * g * s)
+        return {"population_sizes": N, "migration_matrix": epoch.migration_matrix}
+
+    def _delta(self, x, y):
+        if x == y:
+            return 1
+        else:
+            return 0
+
+    # TODO Rename
+    def _get_C_from_N(self, N):
+        """
+        Compute the matrix C, which reprents coalescent probabilities
+        in each population, from Individual effective population sizes.
+        """
+
+        n = len(N)
+        C = np.zeros([n**2, n**2])
+        index_array = np.array(range(n**2)).reshape([n, n])
+        for idx in range(n):
+            C[index_array[idx, idx], index_array[idx, idx]] = 1 / N[idx]
+        return C
+
+    # TODO Rename
+    def _get_G_from_Mp_C(self, Mp, C):
```

Comment:
maybe these two don't need to be separate functions.

Reply:
agreed, I'll probably get rid of the functions to clean it up.